packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,493 @@
+    const assetsWithoutMetadata: CaipAssetType[] = assets.filter(
+      (asset) => !this.state.metadata[asset],
+    );
+
+    // call the snap to get the metadata
+    if (assetsWithoutMetadata.length > 0) {
+      // check if for every asset in assetsWithoutMetadata there is a snap in snaps by chainId else call getAssetSnaps
+      if (
+        !assetsWithoutMetadata.every((asset: CaipAssetType) => {
+          const chainId = parseCaipAssetType(asset);
+          return Boolean(this.#getAssetSnapFor(chainId));
+        })
+      ) {
+        this.#snaps = this.#getAssetSnaps();
+      }
+      await this.#updateAssetsMetadata(assetsWithoutMetadata);
+    }
+  }
+
+  /**
+   * Updates the assets metadata for the given list of assets
+   *
+   * @param assets - The assets to update
+   */
+  async #updateAssetsMetadata(assets: CaipAssetType[]) {
+    // Creates a mapping of scope to their respective assets list.
+    const assetsByScope: Record<CaipChainId, CaipAssetType[]> = {};
+    for (const asset of assets) {
+      const chainId = parseCaipAssetType(asset);
+      if (!assetsByScope[chainId]) {
+        assetsByScope[chainId] = [];
+      }
+      assetsByScope[chainId].push(asset);
+    }
+
+    let newMetadata: Record<CaipAssetType, AssetMetadata>;
+    for (const chainId in assetsByScope) {
+      if (Object.prototype.hasOwnProperty.call(assetsByScope, chainId)) {
+        const assetsForChain = assetsByScope[chainId as CaipChainId];
+        // Now fetch metadata from the associated asset Snaps:
+        const snap = this.#getAssetSnapFor(chainId as CaipChainId);
+        if (snap) {
+          const metadata = await this.#getMetadata(assetsForChain, snap.id);
+          newMetadata = {
+            ...this.state.metadata,
+            ...metadata,
+          };
+        }
+      }
+    }
+    this.update((state) => {
+      state.metadata = newMetadata;
+    });
+  }
+
+  /**
+   * Creates a mapping of CAIP-2 Chain ID to Asset Snaps.
+   *
+   * @returns A mapping of CAIP-2 Chain ID to Asset Snaps.
+   */
+  #getAssetSnaps(): Record<CaipChainId, Snap[]> {
+    const snaps: Record<CaipChainId, Snap[]> = {};
+    const allSnaps = this.#getAllSnaps();
+    const allPermissions = allSnaps.map((snap) =>
+      this.#getSnapsPermissions(snap.id),
+    );
+    // start mock
+    allPermissions.forEach((singlePermission) => {
+      (singlePermission as unknown as AssetEndowment) = {
+        ...singlePermission,
+        'endowment:assets': {
+          scopes: ['bip122:000000000019d6689c085ae165831e93'],
+        },
+      };
+    });
+    (allPermissions[0] as unknown as AssetEndowment) = {
+      ...allPermissions[0],
+      'endowment:assets': {
+        scopes: ['solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1'],
+      },
+    };
+    // end mock
+
+    for (const [index, permission] of allPermissions.entries() as unknown as [
+      number,
+      AssetEndowment,
+    ][]) {
+      const scopes = permission['endowment:assets']?.scopes;
```

Comment:
The scopes are not going to be available in this format since the permission format is different from the manifest format. There is a utility here that can be used to extract the scopes: https://github.com/MetaMask/snaps/blob/362208e725db18baed550ade99087d44e7b537ed/packages/snaps-rpc-methods/src/endowments/name-lookup.ts#L151

Reply:
Thank you! 🙏  Removed permissions mock  from code and updated the format in test here [6cef4ba](https://github.com/MetaMask/core/pull/5138/commits/6cef4ba0aefeae79c3ec7239b029f3a3c6b73031)

Reply:
Resolving this for now, after making the update as suggested in this comment https://github.com/MetaMask/core/pull/5138#discussion_r1939315000 🙏